File: test/engine/plan_generation_test.dart
// RED-then-GREEN unit tests for the Dart plan-generation engine.
// Run: flutter test test/engine/plan_generation_test.dart
//
// Spec under test: lib/engine/plan_generation.dart (single source of truth,
// deterministic training-plan generator, VAL-ONB-056/057).
//
// Determinism contract: identical [PlanIntake] always produces an identical
// [GeneratedPlan] (same exercises, sets, reps, rpe_target, rest, sort order).
// Bodyweight / no-equipment intake yields a non-empty plan with NO ["full_gym"]
// fallback. Contraindicated exercises are excluded.

import 'package:flutter_test/flutter_test.dart';
import 'package:transformfit/engine/plan_generation.dart';

GeneratedPlan _gen(PlanIntake i) => generatePlan(i);

/// Serialize to a stable JSON shape for equality checks.
Map<String, Object?> _json(GeneratedPlan p) => p.toJson();

void main() {
  group('generatePlan: determinism (VAL-ONB-056)', () {
    test('identical intake -> identical plan (byte-equal JSON)', () {
      const intake = PlanIntake(
        goal: 'build_muscle',
        trainingDaysPerWeek: 3,
        equipment: ['dumbbells', 'pull_up_bar'],
        experienceLevel: 'intermediate',
      );
      final a = _gen(intake);
      final b = _gen(intake);
      expect(_json(b), equals(_json(a)));
    });

    test('three repeated runs agree (invariant over run count, red-then-green)', () {
      const intake = PlanIntake(
        goal: 'lose_fat',
        trainingDaysPerWeek: 4,
        equipment: ['kettlebells'],
        experienceLevel: 'beginner',
        limitations: ['knee'],
      );
      final r1 = _json(_gen(intake));
      final r2 = _json(_gen(intake));
      final r3 = _json(_gen(intake));
      expect(r2, equals(r1));
      expect(r3, equals(r1));
    });

    test('deterministic sort order of exercises within a day', () {
      const intake = PlanIntake(
        goal: 'get_fitter',
        trainingDaysPerWeek: 3,
        equipment: ['dumbbells', 'pull_up_bar'],
      );
      final p = _gen(intake);
      expect(p.days, hasLength(3));
      for (final d in p.days) {
        // sortOrder must be 0-based and strictly increasing per day.
        for (int i = 0; i < d.exercises.length; i++) {
          expect(d.exercises[i].sortOrder, equals(i));
        }
        // Exercise ids are unique within a day (round-robin dedup).
        final ids = d.exercises.map((e) => e.id).toList();
        expect(ids.toSet().length, equals(ids.length));
      }
    });
  });

  group('equipment filter (VAL-ONB-026, VAL-ONB-057)', () {
    test('plan uses ONLY selected-equipment + bodyweight movements; never full_gym', () {
      const intake = PlanIntake(
        goal: 'build_muscle',
        trainingDaysPerWeek: 3,
        equipment: ['dumbbells'],
        experienceLevel: 'intermediate',
      );
      final p = _gen(intake);
      final eq = <String>{'bodyweight', ...intake.equipment};
      for (final d in p.days) {
        for (final e in d.exercises) {
          expect(eq.contains(e.equipment), isTrue,
              reason: '${e.id} needs ${e.equipment}, not in $eq');
        }
      }
    });

    test('no-equipment intake -> non-empty plan of bodyweight-only exercises (no full_gym fallback)', () {
      const intake = PlanIntake(
        goal: 'get_fitter',
        trainingDaysPerWeek: 3,
        equipment: <String>[],
        experienceLevel: 'beginner',
      );
      final p = _gen(intake);
      expect(p.days, hasLength(3));
      int total = 0;
      for (final d in p.days) {
        expect(d.exercises, isNotEmpty);
        total += d.exercises.length;
        for (final e in d.exercises) {
          expect(e.equipment, equals('bodyweight'),
              reason:
                  'bodyweight-only intake must NOT fall back to equipment-dependent "${e.id}"');
        }
      }
      expect(total, greaterThan(0));
    });

    test('effectiveEquipment always includes bodyweight (sorted, unique)', () {
      const intake = PlanIntake(
        goal: 'build_strength',
        trainingDaysPerWeek: 4,
        equipment: ['kettlebells', 'dumbbells'],
      );
      final p = _gen(intake);
      expect(p.effectiveEquipment, contains('bodyweight'));
      expect(p.effectiveEquipment, containsAll(<String>['kettlebells', 'dumbbells']));
      // sorted
      final sorted = [...p.effectiveEquipment]..sort();
      expect(p.effectiveEquipment, equals(sorted));
    });
  });

  group('contraindication / injury filter (VAL-ONB-029)', () {
    test('knee limitation excludes squatting/landing movements', () {
      const intake = PlanIntake(
        goal: 'get_fitter',
        trainingDaysPerWeek: 3,
        equipment: ['dumbbells'],
        experienceLevel: 'intermediate',
        limitations: ['knee'],
      );
      final p = _gen(intake);
      final ids = <String>{
        for (final d in p.days) ...d.exercises.map((e) => e.id),
      };
      // Contraindicated for knee: air_squat, bodyweight_lunge, goblet_squat, etc.
      expect(ids.contains('bodyweight_squat'), isFalse);
      expect(ids.contains('bodyweight_lunge'), isFalse);
      expect(ids.contains('goblet_squat'), isFalse);
      // plan should still be non-empty (substitution, not dead-end).
      expect(p.totalExercises, greaterThan(0));
    });

    test('shoulder + wrist limitations remove push-up & OH press family', () {
      const intake = PlanIntake(
        goal: 'build_muscle',
        trainingDaysPerWeek: 3,
        equipment: ['dumbbells', 'barbell', 'bodyweight'],
        experienceLevel: 'intermediate',
        limitations: ['shoulder', 'wrist'],
      );
      final p = _gen(intake);
      final ids = <String>{
        for (final d in p.days) ...d.exercises.map((e) => e.id),
      };
      expect(ids.contains('pushup'), isFalse);
      expect(ids.contains('overhead_press_dumbbell'), isFalse);
      expect(ids.contains('barbell_bench_press'), isFalse);
      expect(ids.contains('dumbbell_bench_press'), isFalse);
      expect(ids.contains('pike_pushup'), isFalse);
      expect(ids.contains('close_grip_pushup'), isFalse);
      expect(ids.contains('tricep_dip'), isFalse);
      expect(p.totalExercises, greaterThan(0));
    });

    test('"none" limitation means no exclusions', () {
      const withNone = PlanIntake(
        goal: 'build_muscle',
        trainingDaysPerWeek: 3,
        equipment: ['dumbbells'],
        experienceLevel: 'intermediate',
        limitations: ['none'],
      );
      const without = PlanIntake(
        goal: 'build_muscle',
        trainingDaysPerWeek: 3,
        equipment: ['dumbbells'],
        experienceLevel: 'intermediate',
      );
      expect(_json(_gen(withNone)), equals(_json(_gen(without))));
    });
  });

  group('volume grid (sets/reps/rpe by goal x experience)', () {
    test('build_muscle / intermediate -> sets=3, reps 8-12, rpe=8', () {
      const intake = PlanIntake(
        goal: 'build_muscle',
        trainingDaysPerWeek: 3,
        equipment: ['dumbbells'],
        experienceLevel: 'intermediate',
      );
      final p = _gen(intake);
      for (final d in p.days) {
        for (final e in d.exercises) {
          expect(e.sets, equals(3));
          expect(e.repsMin, equals(8));
          expect(e.repsMax, equals(12));
          expect(e.rpeTarget, equals(8));
        }
      }
    });

    test('build_strength / advanced -> sets=5, reps 3-5, rpe=9', () {
      const intake = PlanIntake(
        goal: 'build_strength',
        trainingDaysPerWeek: 4,
        equipment: ['barbell'],
        experienceLevel: 'advanced',
      );
      final p = _gen(intake);
      for (final d in p.days) {
        for (final e in d.exercises) {
          expect(e.sets, equals(5));
          expect(e.repsMin, equals(3));
          expect(e.repsMax, equals(5));
          expect(e.rpeTarget, equals(9));
        }
      }
    });

    test('improve_mobility / beginner -> sets=2, reps 8-12', () {
      const intake = PlanIntake(
        goal: 'improve_mobility',
        trainingDaysPerWeek: 3,
        equipment: <String>[],
        experienceLevel: 'beginner',
      );
      final p = _gen(intake);
      for (final d in p.days) {
        for (final e in d.exercises) {
          expect(e.sets, equals(2));
          expect(e.repsMin, equals(8));
          expect(e.repsMax, equals(12));
          expect(e.rpeTarget, equals(6));
        }
      }
    });

    test('null experienceLevel falls back to intermediate row', () {
      const intakeNull = PlanIntake(
        goal: 'lose_fat',
        trainingDaysPerWeek: 3,
        equipment: ['dumbbells'],
        experienceLevel: null,
      );
      const intakeInter = PlanIntake(
        goal: 'lose_fat',
        trainingDaysPerWeek: 3,
        equipment: ['dumbbells'],
        experienceLevel: 'intermediate',
      );
      expect(_json(_gen(intakeNull)), equals(_json(_gen(intakeInter))));
    });

    test('rest seconds: compound=120, isolation=60 (non-beginner)', () {
      const intake = PlanIntake(
        goal: 'build_muscle',
        trainingDaysPerWeek: 3,
        equipment: ['dumbbells'],
        experienceLevel: 'intermediate',
      );
      final p = _gen(intake);
      for (final d in p.days) {
        for (final e in d.exercises) {
          if (e.isCompound) {
            expect(e.restSeconds, equals(120));
          } else {
            expect(e.restSeconds, equals(60));
          }
        }
      }
    });
  });

  group('split by training days per week (cadence, VAL-ONB-027)', () {
    test('3 days -> 3 plan days, full body split codes', () {
      final p = _gen(const PlanIntake(
        goal: 'get_fitter',
        trainingDaysPerWeek: 3,
        equipment: ['dumbbells'],
      ));
      expect(p.days, hasLength(3));
      expect(p.days.map((d) => d.split).toSet(), equals(<String>{'full'}));
    });

    test('4 days -> 4 plan days, upper/lower', () {
      final p = _gen(const PlanIntake(
        goal: 'build_muscle',
        trainingDaysPerWeek: 4,
        equipment: ['dumbbells'],
      ));
      expect(p.days, hasLength(4));
      expect(p.days[0].split, equals('upper'));
      expect(p.days[1].split, equals('lower'));
      expect(p.days[2].split, equals('upper'));
      expect(p.days[3].split, equals('lower'));
    });

    test('5 days -> 5 plan days, push/pull/legs/upper/lower', () {
      final p = _gen(const PlanIntake(
        goal: 'get_fitter',
        trainingDaysPerWeek: 5,
        equipment: ['dumbbells', 'barbell'],
      ));
      expect(p.days, hasLength(5));
      expect(p.days[0].split, equals('push'));
      expect(p.days[1].split, equals('pull'));
      expect(p.days[2].split, equals('legs'));
      expect(p.days[3].split, equals('upper'));
      expect(p.days[4].split, equals('lower'));
      // target exercises/day == 4 for 5 days
      for (final d in p.days) {
        expect(d.exercises.length, lessThanOrEqualTo(4));
      }
    });

    test('2 days -> 2 full-body days', () {
      final p = _gen(const PlanIntake(
        goal: 'build_strength',
        trainingDaysPerWeek: 2,
        equipment: ['barbell'],
      ));
      expect(p.days, hasLength(2));
      // target exercises/day == 6 for 2 days
      for (final d in p.days) {
        expect(d.exercises.length, lessThanOrEqualTo(6));
      }
    });
  });

  group('consequential behavior (VAL-ONB-050): changing an answer changes the plan', () {
    test('changing equipment changes the exercise set', () {
      const a = PlanIntake(
        goal: 'get_fitter',
        trainingDaysPerWeek: 3,
        equipment: ['dumbbells'],
      );
      const b = PlanIntake(
        goal: 'get_fitter',
        trainingDaysPerWeek: 3,
        equipment: ['barbell'],
      );
      expect(_json(_gen(b)) == _json(_gen(a)), isFalse);
    });

    test('changing goal changes sets/reps/rpe', () {
      const a = PlanIntake(
        goal: 'build_muscle',
        trainingDaysPerWeek: 3,
        equipment: ['dumbbells'],
        experienceLevel: 'intermediate',
      );
      const b = PlanIntake(
        goal: 'build_strength',
        trainingDaysPerWeek: 3,
        equipment: ['dumbbells'],
        experienceLevel: 'intermediate',
      );
      expect(_json(_gen(b)) == _json(_gen(a)), isFalse);
    });

    test('adding a limitation reduces or changes the exercise set, never errors', () {
      const a = PlanIntake(
        goal: 'get_fitter',
        trainingDaysPerWeek: 3,
        equipment: ['dumbbells'],
      );
      final b = PlanIntake(
        goal: 'get_fitter',
        trainingDaysPerWeek: 3,
        equipment: ['dumbbells'],
        limitations: ['back', 'knee', 'shoulder'],
      );
      final pa = _gen(a);
      final pb = _gen(b);
      // With many limitations the bodyweight-only plan body gets smaller
      // but is NEVER empty.
      expect(pb.totalExercises, greaterThan(0));
      expect(pb.totalExercises, lessThanOrEqualTo(pa.totalExercises));
    });
  });

  group('output contract', () {
    test('every exercise has non-empty name, positive sets/reps', () {
      const intake = PlanIntake(
        goal: 'build_muscle',
        trainingDaysPerWeek: 5,
        equipment: ['dumbbells', 'pull_up_bar'],
        experienceLevel: 'advanced',
      );
      final p = _gen(intake);
      for (final d in p.days) {
        for (final e in d.exercises) {
          expect(e.name, isNotEmpty);
          expect(e.muscleGroup, isNotEmpty);
          expect(e.equipment, isNotEmpty);
          expect(e.sets, greaterThan(0));
          expect(e.repsMax, greaterThanOrEqualTo(e.repsMin));
          expect(e.rpeTarget, inInclusiveRange(1, 10));
          expect(e.restSeconds, greaterThan(0));
        }
      }
    });
  });
}

File: supabase/functions/_shared/engines/plan_generation_test.ts
// Deno unit tests for the Deno plan-generation engine — mirror of the Dart
// tests (test/engine/plan_generation_test.dart). Run:
//   deno test --allow-all supabase/functions/_shared/engines/plan_generation_test.ts
//
// Single spec with the Dart engine (deterministic training-plan generator,
// VAL-ONB-056/057). Identical intake -> Identical plan. Bodyweight/no-equipment
// intake yields a non-empty plan with no full_gym fallback. Contraindicated
// exercises excluded.
import { assert, assertEquals, assertNotEquals } from "@std/assert";
import { generatePlan, type PlanIntake } from "./plan_generation.ts";

const _gen = (i: PlanIntake) => generatePlan(i);
const _json = (p: ReturnType<typeof generatePlan>) => JSON.stringify(p);

Deno.test("determinism: identical intake -> byte-equal JSON", () => {
  const intake: PlanIntake = {
    goal: "build_muscle",
    trainingDaysPerWeek: 3,
    equipment: ["dumbbells", "pull_up_bar"],
    experienceLevel: "intermediate",
  };
  assertEquals(_json(_gen(intake)), _json(_gen(intake)));
  assertEquals(_json(_gen(intake)), _json(_gen(intake)));
});

Deno.test("determinism: repeated runs on a more complex intake agree", () => {
  const intake: PlanIntake = {
    goal: "lose_fat",
    trainingDaysPerWeek: 4,
    equipment: ["kettlebells"],
    experienceLevel: "beginner",
    limitations: ["knee"],
  };
  const r1 = _json(_gen(intake));
  const r2 = _json(_gen(intake));
  const r3 = _json(_gen(intake));
  assertEquals(r2, r1);
  assertEquals(r3, r1);
});

Deno.test("deterministic sort order and unique ids per day", () => {
  const intake: PlanIntake = {
    goal: "get_fitter",
    trainingDaysPerWeek: 3,
    equipment: ["dumbbells", "pull_up_bar"],
  };
  const p = _gen(intake);
  assertEquals(p.days.length, 3);
  for (const d of p.days) {
    for (let i = 0; i < d.exercises.length; i++) {
      assertEquals(d.exercises[i].sortOrder, i);
    }
    const ids = d.exercises.map((e) => e.id);
    assertEquals(new Set(ids).size, ids.length);
  }
});

Deno.test("equipment filter: uses only selected + bodyweight, never full_gym", () => {
  const intake: PlanIntake = {
    goal: "build_muscle",
    trainingDaysPerWeek: 3,
    equipment: ["dumbbells"],
    experienceLevel: "intermediate",
  };
  const p = _gen(intake);
  const eq = new Set<string>(["bodyweight", ...intake.equipment]);
  for (const d of p.days) {
    for (const e of d.exercises) {
      assert(
        eq.has(e.equipment),
        `${e.id} needs ${e.equipment}, not in ${[...eq]}`,
      );
    }
  }
});

Deno.test("no-equipment intake -> non-empty bodyweight-only plan (no full_gym fallback)", () => {
  const intake: PlanIntake = {
    goal: "get_fitter",
    trainingDaysPerWeek: 3,
    equipment: [],
    experienceLevel: "beginner",
  };
  const p = _gen(intake);
  assertEquals(p.days.length, 3);
  let total = 0;
  for (const d of p.days) {
    assert(d.exercises.length > 0);
    total += d.exercises.length;
    for (const e of d.exercises) {
      assertEquals(
        e.equipment,
        "bodyweight",
        `bodyweight-only intake must not fall back to "${e.id}" using ${e.equipment}`,
      );
    }
  }
  assert(total > 0);
});

Deno.test("effectiveEquipment includes bodyweight, sorted unique", () => {
  const intake: PlanIntake = {
    goal: "build_strength",
    trainingDaysPerWeek: 4,
    equipment: ["kettlebells", "dumbbells"],
  };
  const p = _gen(intake);
  assert(p.effectiveEquipment.includes("bodyweight"));
  assert(p.effectiveEquipment.includes("kettlebells"));
  assert(p.effectiveEquipment.includes("dumbbells"));
  assertEquals([...p.effectiveEquipment].sort(), p.effectiveEquipment);
});

Deno.test("knee limitation excludes squatting movements", () => {
  const intake: PlanIntake = {
    goal: "get_fitter",
    trainingDaysPerWeek: 3,
    equipment: ["dumbbells"],
    experienceLevel: "intermediate",
    limitations: ["knee"],
  };
  const p = _gen(intake);
  const ids = new Set(p.days.flatMap((d) => d.exercises.map((e) => e.id)));
  assert(!ids.has("bodyweight_squat"));
  assert(!ids.has("bodyweight_lunge"));
  assert(!ids.has("goblet_squat"));
  assert(p.days.reduce((s, d) => s + d.exercises.length, 0) > 0);
});

Deno.test("shoulder+wrist limitations remove push-up & OH press family", () => {
  const intake: PlanIntake = {
    goal: "build_muscle",
    trainingDaysPerWeek: 3,
    equipment: ["dumbbells", "barbell", "bodyweight"],
    experienceLevel: "intermediate",
    limitations: ["shoulder", "wrist"],
  };
  const p = _gen(intake);
  const ids = new Set(p.days.flatMap((d) => d.exercises.map((e) => e.id)));
  assert(!ids.has("pushup"));
  assert(!ids.has("overhead_press_dumbbell"));
  assert(!ids.has("barbell_bench_press"));
  assert(!ids.has("dumbbell_bench_press"));
  assert(!ids.has("pike_pushup"));
  assert(!ids.has("close_grip_pushup"));
  assert(!ids.has("tricep_dip"));
  assert(p.days.reduce((s, d) => s + d.exercises.length, 0) > 0);
});

Deno.test('"none" limitation means no exclusions', () => {
  const withNone: PlanIntake = {
    goal: "build_muscle",
    trainingDaysPerWeek: 3,
    equipment: ["dumbbells"],
    experienceLevel: "intermediate",
    limitations: ["none"],
  };
  const without: PlanIntake = {
    goal: "build_muscle",
    trainingDaysPerWeek: 3,
    equipment: ["dumbbells"],
    experienceLevel: "intermediate",
  };
  assertEquals(_json(_gen(withNone)), _json(_gen(without)));
});

Deno.test("volume grid: build_muscle/intermediate -> 3x8-12@8", () => {
  const p = _gen({
    goal: "build_muscle",
    trainingDaysPerWeek: 3,
    equipment: ["dumbbells"],
    experienceLevel: "intermediate",
  });
  for (const d of p.days) {
    for (const e of d.exercises) {
      assertEquals(e.sets, 3);
      assertEquals(e.repsMin, 8);
      assertEquals(e.repsMax, 12);
      assertEquals(e.rpeTarget, 8);
    }
  }
});

Deno.test("volume grid: build_strength/advanced -> 5x3-5@9", () => {
  const p = _gen({
    goal: "build_strength",
    trainingDaysPerWeek: 4,
    equipment: ["barbell"],
    experienceLevel: "advanced",
  });
  for (const d of p.days) {
    for (const e of d.exercises) {
      assertEquals(e.sets, 5);
      assertEquals(e.repsMin, 3);
      assertEquals(e.repsMax, 5);
      assertEquals(e.rpeTarget, 9);
    }
  }
});

Deno.test("null experienceLevel falls back to intermediate row (plan equals explicit intermediate)", () => {
  const a: PlanIntake = {
    goal: "lose_fat",
    trainingDaysPerWeek: 3,
    equipment: ["dumbbells"],
    experienceLevel: null,
  };
  const b: PlanIntake = {
    goal: "lose_fat",
    trainingDaysPerWeek: 3,
    equipment: ["dumbbells"],
    experienceLevel: "intermediate",
  };
  assertEquals(_json(_gen(a)), _json(_gen(b)));
});

Deno.test("rest seconds: compound=120, isolation=60 (non-beginner)", () => {
  const p = _gen({
    goal: "build_muscle",
    trainingDaysPerWeek: 3,
    equipment: ["dumbbells"],
    experienceLevel: "intermediate",
  });
  for (const d of p.days) {
    for (const e of d.exercises) {
      assertEquals(e.restSeconds, e.isCompound ? 120 : 60);
    }
  }
});

Deno.test("split: 3 days -> 3 full-body days", () => {
  const p = _gen({
    goal: "get_fitter",
    trainingDaysPerWeek: 3,
    equipment: ["dumbbells"],
  });
  assertEquals(p.days.length, 3);
  assertEquals(new Set(p.days.map((d) => d.split)), new Set(["full"]));
});

Deno.test("split: 4 days -> upper/lower/upper/lower", () => {
  const p = _gen({
    goal: "build_muscle",
    trainingDaysPerWeek: 4,
    equipment: ["dumbbells"],
  });
  assertEquals(p.days.length, 4);
  assertEquals(p.days[0].split, "upper");
  assertEquals(p.days[1].split, "lower");
  assertEquals(p.days[2].split, "upper");
  assertEquals(p.days[3].split, "lower");
});

Deno.test("split: 5 days -> push/pull/legs/upper/lower", () => {
  const p = _gen({
    goal: "get_fitter",
    trainingDaysPerWeek: 5,
    equipment: ["dumbbells", "barbell"],
  });
  assertEquals(p.days.length, 5);
  assertEquals(p.days[0].split, "push");
  assertEquals(p.days[1].split, "pull");
  assertEquals(p.days[2].split, "legs");
  assertEquals(p.days[3].split, "upper");
  assertEquals(p.days[4].split, "lower");
  for (const d of p.days) assert(d.exercises.length <= 4);
});

Deno.test("split: 2 days -> 2 full-body days", () => {
  const p = _gen({
    goal: "build_strength",
    trainingDaysPerWeek: 2,
    equipment: ["barbell"],
  });
  assertEquals(p.days.length, 2);
  for (const d of p.days) assert(d.exercises.length <= 6);
});

Deno.test("consequential: changing equipment changes the plan", () => {
  const a: PlanIntake = {
    goal: "get_fitter",
    trainingDaysPerWeek: 3,
    equipment: ["dumbbells"],
  };
  const b: PlanIntake = {
    goal: "get_fitter",
    trainingDaysPerWeek: 3,
    equipment: ["barbell"],
  };
  assertNotEquals(_json(_gen(a)), _json(_gen(b)));
});

Deno.test("consequential: changing goal changes sets/reps/rpe", () => {
  const a: PlanIntake = {
    goal: "build_muscle",
    trainingDaysPerWeek: 3,
    equipment: ["dumbbells"],
    experienceLevel: "intermediate",
  };
  const b: PlanIntake = {
    goal: "build_strength",
    trainingDaysPerWeek: 3,
    equipment: ["dumbbells"],
    experienceLevel: "intermediate",
  };
  assertNotEquals(_json(_gen(a)), _json(_gen(b)));
});

Deno.test("output contract: positive sets, repsMax>=repsMin, rpe 1..10", () => {
  const p = _gen({
    goal: "build_muscle",
    trainingDaysPerWeek: 5,
    equipment: ["dumbbells", "pull_up_bar"],
    experienceLevel: "advanced",
  });
  for (const d of p.days) {
    for (const e of d.exercises) {
      assert(e.name.length > 0);
      assert(e.sets > 0);
      assert(e.repsMax >= e.repsMin);
      assert(e.rpeTarget >= 1 && e.rpeTarget <= 10);
      assert(e.restSeconds > 0);
    }
  }
});
